eth/filters/filter.go

```diff
@@ -95,109 +95,155 @@ func newFilter(sys *FilterSystem, addresses []common.Address, topics [][]common.
+		if f.block != nil {
+			header, err := f.sys.backend.HeaderByHash(ctx, *f.block)
+			if err != nil {
+				errChan <- err
+				return
+			}
+			if header == nil {
+				errChan <- errors.New("unknown block")
+				return
+			}
+			logs, err := f.blockLogs(ctx, header)
+			if err != nil {
+				errChan <- err
+				return
 			}
-		case rpc.SafeBlockNumber.Int64():
-			hdr, _ = f.sys.backend.HeaderByNumber(ctx, rpc.SafeBlockNumber)
-			if hdr == nil {
-				return 0, errors.New("safe header not found")
+			for _, log := range logs {
+				logChan <- log
 			}
-		default:
-			return number, nil
+			return
 		}
-		return hdr.Number.Int64(), nil
-	}
-	if f.begin, err = resolveSpecial(f.begin); err != nil {
-		return nil, err
-	}
-	if f.end, err = resolveSpecial(f.end); err != nil {
-		return nil, err
-	}
-	// Gather all indexed logs, and finish with non indexed ones
-	var (
-		logs           []*types.Log
-		end            = uint64(f.end)
-		size, sections = f.sys.backend.BloomStatus()
-	)
-	if indexed := sections * size; indexed > uint64(f.begin) {
-		if indexed > end {
-			logs, err = f.indexedLogs(ctx, end)
-		} else {
-			logs, err = f.indexedLogs(ctx, indexed-1)
+		// Short-cut if all we care about is pending logs
+		if f.begin == rpc.PendingBlockNumber.Int64() {
+			if f.end != rpc.PendingBlockNumber.Int64() {
+				errChan <- errors.New("invalid block range")
+				return
+			}
+			for _, log := range f.pendingLogs() {
+				logChan <- log
+			}
+			return
 		}
-		if err != nil {
-			return logs, err
+		// Figure out the limits of the filter range
+		header, _ := f.sys.backend.HeaderByNumber(ctx, rpc.LatestBlockNumber)
+		if header == nil {
+			errChan <- nil
+			return
 		}
-	}
-	rest, err := f.unindexedLogs(ctx, end)
-	logs = append(logs, rest...)
-	if pending {
-		pendingLogs, err := f.pendingLogs()
+		var (
+			err     error
+			head    = header.Number.Int64()
+			pending = f.end == rpc.PendingBlockNumber.Int64()
+		)
+		resolveSpecial := func(number int64) (int64, error) {
+			var hdr *types.Header
+			switch number {
+			case rpc.LatestBlockNumber.Int64():
+				return head, nil
+			case rpc.PendingBlockNumber.Int64():
+				// we should return head here since we've already captured
+				// that we need to get the pending logs in the pending boolean above
+				return head, nil
+			case rpc.FinalizedBlockNumber.Int64():
+				hdr, _ = f.sys.backend.HeaderByNumber(ctx, rpc.FinalizedBlockNumber)
+				if hdr == nil {
+					return 0, errors.New("finalized header not found")
+				}
+			case rpc.SafeBlockNumber.Int64():
+				hdr, _ = f.sys.backend.HeaderByNumber(ctx, rpc.SafeBlockNumber)
+				if hdr == nil {
+					return 0, errors.New("safe header not found")
+				}
+			default:
+				return number, nil
+			}
+			return hdr.Number.Int64(), nil
+		}
+		if f.begin, err = resolveSpecial(f.begin); err != nil {
+			errChan <- err
+			return
+		}
+		if f.end, err = resolveSpecial(f.end); err != nil {
+			errChan <- err
+			return
+		}
```

Comment:
It's kind of wrong to resolve the `f.begin ... f.end` range in the goroutine. They should be resolved at the start of processing the logs request. It doesn't matter too much right now, since the async processing is really only used once in `Logs`. However, it will start mattering when `logsAsync` is used inside of the subscription. Since some of the block specifiers can change over time ("safe", "finalized", "latest"), invoking `logsAsync` multiple times could produce incomplete results.

Reply:
thanks for the input